src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -19,22 +23,158 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 *
 	 * @return boolean Whether an element was found.
 	 */
-	public function step( $insertion_mode = 'in-body' ) {
-		switch ( $insertion_mode ) {
+	public function step( $insertion_mode = null ) {
+		switch ( $insertion_mode ?: $this->insertion_mode ) {
 			case 'in-body':
 				return $this->step_in_body();
 
 			default:
-				return false;
+				return self::NOT_IMPLEMENTED_YET;
 		}
 	}
 
+	/**
+	 * Parses next element in the 'in head' insertion mode.
+	 *
+	 * Not yet implemented.
+	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inhead
+	 *
+	 * @return false
+	 */
+	private function step_in_head() {
+		return self::NOT_IMPLEMENTED_YET;
+	}
+
 	/**
 	 * Parses next element in the 'in body' insertion mode.
 	 *
+	 * @see https://html.spec.whatwg.org/#parsing-main-inbody
+	 *
 	 * @return boolean Whether an element was found.
 	 */
 	private function step_in_body() {
+		ignored:
+		parent::set_bookmark( 'current' );
+		if ( ! $this->next_tag( self::$query ) ) {
+			return false;
+		}
+
+		$tag_name = $this->get_tag();
+		$tag_type = $this->is_tag_closer() ? 'closer' : 'opener';
+
+		/*
+		 * > A start tag whose tag name is "html"
+		 */
+		if ( 'HTML' === $tag_name && 'opener' === $tag_type ) {
+			goto ignored;
+		}
+
+		/*
+		 * > A start tag whose tag name is one of: "base", "basefont", "bgsound",
+		 * > "link", "meta", "noframes", "script", "style", "template", "title"
+		 *
+		 * > An end tag whose tag name is "template"
+		 */
+		if (
+			'opener' === $tag_type && (
+				'BASE' === $tag_name ||
+				'BASEFONT' === $tag_name ||
+				'BGSOUND' === $tag_name ||
+				'LINK' === $tag_name ||
+				'META' === $tag_name ||
+				'NOFRAMES' === $tag_name ||
+				'SCRIPT' === $tag_name ||
+				'STYLE' === $tag_name ||
+				'TEMPLATE' === $tag_name ||
+				'TITLE' === $tag_name
+			) ||
+			(
+				'closer' === $tag_type &&
```

Comment:
This allows for more brevity at the expense of structuring the code differently than the specification. Subtle issues aside, this requires the reader translate between the spec and this function. I know there's plenty of space to make this code more succinct than the specification, but I would keep the structure the same. It will be of so much help once we have to deal with a case that doesn't behave as expected.

Reply:
actually this _is_ how I've structured my code, to follow the spec, and I've included the appropriate quotes at the start of each section. I was playing with how to match it.

wanted to link to the individual parsing steps but they aren't given anchor ids, so the quotes are what we have to link to them.

I was playing with ideas on how best to corroborate with the spec.